### ISSUE TITLE
(Map k v) and [(k, v)] are encoded in opposite directions

### ISSUE THREAD

**Issue by sjakobi** [fpco/store#97]
```
λ> let w10 = Prelude.map (\x -> (x, x)) [1..10 :: Word8]
λ> let w10m = Map.fromList w10
λ> encode w10
"\n\NUL\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\STX\STX\ETX\ETX\EOT\EOT\ENQ\ENQ\ACK\ACK\a\a\b\b\t\t\n\n"
λ> encode w10m
"\n\NUL\NUL\NUL\NUL\NUL\NUL\NUL\n\n\t\t\b\b\a\a\ACK\ACK\ENQ\ENQ\EOT\EOT\ETX\ETX\STX\STX\SOH\SOH"
```

Not sure if it's a bug but it's certainly surprising. It also seems to me that this should cause a bug in #96 where I use `fromDistinctAscList` although the input is strictly descending – but the only test failure seems unrelated.

**Comment by sjakobi** [comment]
A possible fix is to replace [the current implementation of `pokeMap`](https://github.com/fpco/store/blob/b611f6f0d5a5059e0c052f2db82e40e1206aac63/src/Data/Store/Internal.hs#L219) with `pokeSequence . mapToList`.

**Comment by sjakobi** [comment]
> It also seems to me that this should cause a bug in #96  where I use fromDistinctAscList although the input is strictly descending – but the only test failure seems unrelated.

Ok, so my new implementation of `peek` via `fromDistinctAscList` is definitely affected by the descending order of the elements in the serialized bytestring:

```
λ> let m = fromList $ zip [1 .. 10 :: Word8] (repeat ())
λ> let m' = decodeEx $ encode m :: Map Word8 ()
λ> m'
fromList [(10,()),(9,()),(8,()),(7,()),(6,()),(5,()),(4,()),(3,()),(2,()),(1,())]
λ> m
fromList [(1,()),(2,()),(3,()),(4,()),(5,()),(6,()),(7,()),(8,()),(9,()),(10,())]
λ> m == m'
False
```

The strange thing is that the tests that roundtrip `Map Int8 Int8` and `Map Int64 Int64` still don't fail.

One thing that I noticed is that `store-test` seems to ignore the `--depth` arguments that I can pass in on the command line. `stack test :store-test --test-arguments "--match Map --depth 20"` finishes in about the same time as `--depth 1` ( < 20 ms).

Because the `Serial` instance for `Map` only starts to generate maps with multiple elements for depth >= 3, I think the depth is somehow limited to <= 2.

**Comment by sjakobi** [comment]
> Because the Serial instance for Map only starts to generate maps with multiple elements for depth >= 3, I think the depth is somehow limited to <= 2.

Right, here's [the culprit](https://github.com/fpco/store/blob/b611f6f0d5a5059e0c052f2db82e40e1206aac63/test/Data/StoreSpec.hs#L310)! ;)

If I change the depth to 3, I get 

```
    Roundtrips (Data.Map.Base.Map GHC.Int.Int8 GHC.Int.Int8) FAILED [1]
    Roundtrips (Data.Map.Base.Map GHC.Int.Int64 GHC.Int.Int64) FAILED [2]
```

**Comment by mgsloan** [comment]
Doh!  Outputting the map backwards is quite a screwup..  I'm wary that if we switch to fromDistinctAscList, folks might run into some funky behavior if loading data from an old store version with a new store version.

One possibility for poking maps could actually just be `poke (olength x) >> omapM_ poke x`

While it's a bit ugly, what if it inserted and checked a magic Word8 or Word32?  This'd at least cause it to usually be an error, rather than constructing an invalid Map.  We'd also certainly want to be a major version number bump.

**Comment by sjakobi** [comment]
> poke (olength x) >> omapM_ poke x

I think that would only poke the values, not the keys. It turns out that `pokeSequence . mapToList` already performs better than the old implementation but I'm sure there's still room for improvement.
What I like about this implementation is that there's little room for error and it's pretty DRY.

> While it's a bit ugly, what if it inserted and checked a magic Word8 or Word32? This'd at least cause it to usually be an error, rather than constructing an invalid Map.

I think that's a good idea, I have implemented it in my new PR #101.
